packages/@aws-cdk/aws-codepipeline-actions/lib/cloudformation/stackset-action.ts

```diff
@@ -0,0 +1,691 @@
+   * If you don't, the CodePipeline Construct will create new Stacks in your CDK app containing those buckets,
+   * that you will need to `cdk deploy` before deploying the main, Pipeline-containing Stack.
+   *
+   * @default - the Action resides in the same region as the Pipeline
+   */
+  readonly region?: string;
+}
+
+/**
+ * Where Stack Instances will be created from the StackSet
+ */
+export abstract class StackInstances {
+  /**
+   * Create stack instances in a literal set of accounts or organizational units, and a set of regions
+   *
+   * Stack Instances will be created in every combination of region and account, or region and
+   * Organizational Units (OUs).
+   *
+   * If this is set of Organizational Units, you must have selected `StackSetDeploymentModel.organizations()`
+   * as deployment model.
+   */
+  public static fromList(targets: string[], regions: string[]): StackInstances {
+    if (targets.length === 0) {
+      throw new Error("'targets' may not be an empty list");
+    }
+
+    if (regions.length === 0) {
+      throw new Error("'regions' may not be an empty list");
+    }
+
+    return new class extends StackInstances {
+      public bind(_scope: Construct): StackInstancesBindResult {
+        return {
+          stackSetConfiguration: {
+            DeploymentTargets: targets.join(','),
+            Regions: regions.join(','),
+          },
+        };
+      }
+    }();
+  }
+
+  /**
+   * Create stack instances in a set of accounts or organizational units taken from the pipeline artifacts, and a set of regions
+   *
+   * Stack Instances will be created in every combination of region and account, or region and
+   * Organizational Units (OUs).
+   *
+   * If this is set of Organizational Units, you must have selected `StackSetDeploymentModel.organizations()`
+   * as deployment model.
+   */
+  public static fromArtifact(artifactPath: codepipeline.ArtifactPath, regions: string[]): StackInstances {
+    if (regions.length === 0) {
+      throw new Error("'regions' may not be an empty list");
+    }
+
+    return new class extends StackInstances {
+      public readonly artifactsReferenced?: codepipeline.Artifact[] | undefined = [artifactPath.artifact];
+      public bind(_scope: Construct): StackInstancesBindResult {
+        return {
+          stackSetConfiguration: {
+            DeploymentTargets: artifactPath.location,
+            Regions: regions.join(','),
+          },
+        };
+      }
+    }();
+  }
+
+  /**
+   * The artifacts referenced by the properties of this deployment target
+   *
+   * Does not need to be called by app builders.
+   */
+  readonly artifactsReferenced?: codepipeline.Artifact[];
+
+  /**
+   * Called to attach the stack set instances to a stackset action
+   *
+   * Does not need to be called by app builders.
+   */
+  public abstract bind(scope: Construct): StackInstancesBindResult;
```

Comment:
Can we make this `@internal`?

Reply:
We *can*. I'm not exactly keen on it. It artificially closes off the API from other implementors, and makes the framework less extensible. Is it worth it?

Reply:
But the chances of anyone actually needing to implement this are so tiny, right?

Unlike with something like a CodePipeline Action, that could _easily_ be required to have a customer implementation (and in fact, I know they already do).

So, i think a universal rule here doesn't work - it's a judgement call each time on the side of the API author.